### PR TITLE
[db migration] change datasources-clusters foreign key to cluster_id

### PR DESCRIPTION
**Pull request by serenajiang**
### CATEGORY
Choose one

- [ ] Bug Fix
- [ ] Enhancement (new features, refinement)
- [ ] Refactor
- [ ] Add tests
- [ ] Build / Development Environment
- [ ] Documentation
- [X] Schema change

### SUMMARY
Change the foreign key for `datasources`->`clusters` from `cluster_name` to `cluster_id` (the primary key of `clusters`).

### TEST PLAN
`tox` passes.

### ADDITIONAL INFORMATION

- [ ] Has associated issue:
- [ ] Changes UI
- [X] Requires DB Migration.
- [X] Confirm DB Migration upgrade and downgrade tested.
- [ ] Introduces new feature or API
- [ ] Removes existing feature or API

### REVIEWERS
@john-bodley @mistercrunch

### CODE REVIEW

File: superset/connectors/druid/models.py [hunk 1]
```diff
@@ -608,17 +614,13 @@ def lookup_datasource(d: DruidDatasource) -> Optional[DruidDatasource]:
...
                 .first()
             )
 
         def lookup_cluster(d: DruidDatasource) -> Optional[DruidCluster]:
-            return (
-                db.session.query(DruidCluster)
-                .filter_by(cluster_name=d.cluster_name)
-                .one()
-            )
+            return db.session.query(DruidCluster).filter_by(id=d.cluster_id).one()
```
Comment (john-bodley) [inline superset/connectors/druid/models.py]: Somewhat unrelated to this PR but the return type here is `Optional[DruidCluster]` whereas the call is to `.one()` (which could throw) as opposed to `.one_or_none()`.

File: superset/connectors/druid/models.py [hunk 2]
```diff
@@ -1619,12 +1621,7 @@ def get_having_filters(self, raw_filters: List[Dict]) -> "Having":
     def query_datasources_by_name(
         cls, session: Session, database: Database, datasource_name: str, schema=None
     ) -> List["DruidDatasource"]:
-        return (
-            session.query(cls)
-            .filter_by(cluster_name=database.id)
-            .filter_by(datasource_name=datasource_name)
-            .all()
-        )
+        return session.query(cls).filter_by(datasource_name=datasource_name).all()
```
Comment (john-bodley) [inline superset/connectors/druid/models.py]: Shouldn't there be a filter on the cluster as well?

File: superset/connectors/druid/views.py [hunk 3]
```diff
@@ -341,7 +341,7 @@ def pre_add(self, datasource):
         with db.session.no_autoflush:
             query = db.session.query(models.DruidDatasource).filter(
                 models.DruidDatasource.datasource_name == datasource.datasource_name,
-                models.DruidDatasource.cluster_name == datasource.cluster.id,
+                models.DruidDatasource.cluster_id == datasource.cluster_id,
```
Comment (john-bodley) [inline superset/connectors/druid/views.py]: Was this simply wrong before?

File: superset/migrations/versions/e96dbf2cfef0_datasource_cluster_fk.py [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+        )
+        uq_constraint_name = generic_find_uq_constraint_name(
+            "datasources", {"cluster_name", "datasource_name"}, insp
+        )
+
+        batch_op.alter_column("cluster_id", existing_type=sa.Integer, nullable=False)
+        batch_op.drop_constraint(fk_constraint_name, type_="foreignkey")
+        batch_op.drop_constraint(uq_constraint_name, type_="unique")
+        batch_op.drop_column("cluster_name")
+        batch_op.create_unique_constraint(
```
Comment (john-bodley) [inline superset/migrations/versions/e96dbf2cfef0_datasource_cluster_fk.py]: The comment on line #59 reads as it pertains to the entire with statement. Maybe that should appear on like #61 and a similar comment be added on line #72 with regards to adding the new constraint and foreign key.

File: superset/migrations/versions/e96dbf2cfef0_datasource_cluster_fk.py [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+
+    statement = datasources.update().values(
+        cluster_name=sa.select([clusters.c.cluster_name])
+        .where(datasources.c.cluster_id == clusters.c.id)
+        .as_scalar()
+    )
+    db.session.execute(statement)
+    db.session.commit()
+
+    # Drop new column and re-add constraints
```
Comment (john-bodley) [inline superset/migrations/versions/e96dbf2cfef0_datasource_cluster_fk.py]: See ^^.

File: superset/utils/core.py [hunk 6]
```diff
@@ -485,7 +486,9 @@ def readfile(file_path: str) -> Optional[str]:
     return content
 
 
-def generic_find_constraint_name(table, columns, referenced, db):
+def generic_find_constraint_name(
```
Comment (john-bodley) [inline superset/utils/core.py]: 👍

File: superset/utils/core.py [hunk 7]
```diff
@@ -494,7 +497,9 @@ def generic_find_constraint_name(table, columns, referenced, db):
             return fk.name
 
 
-def generic_find_fk_constraint_name(table, columns, referenced, insp):
+def generic_find_fk_constraint_name(
```
Comment (john-bodley) [inline superset/utils/core.py]: 👍

File: tests/dict_import_export_tests.py [hunk 8]
```diff
@@ -87,11 +92,20 @@ def create_table(self, name, schema="", id=0, cols_names=[], metric_names=[]):
         return table, dict_rep
 
     def create_druid_datasource(self, name, id=0, cols_names=[], metric_names=[]):
-        name = "{0}{1}".format(NAME_PREFIX, name)
         cluster_name = "druid_test"
+        cluster = (
```
Comment (john-bodley) [inline tests/dict_import_export_tests.py]: This is repeated below. Can we create a utility method instead.

File: superset/connectors/druid/views.py (hunk 3 above)
Comment (serenajiang) [reply superset/connectors/druid/views.py]: Yep 🤷‍♀

File: tests/dict_import_export_tests.py (hunk 8 above)
Comment (serenajiang) [reply tests/dict_import_export_tests.py]: Found one that existed 🎉

File: superset/connectors/druid/models.py [hunk 9]
```diff
@@ -1621,8 +1623,9 @@ def query_datasources_by_name(
     ) -> List["DruidDatasource"]:
         return (
             session.query(cls)
-            .filter_by(cluster_name=database.id)
-            .filter_by(datasource_name=datasource_name)
+            .join(DruidCluster)
+            .filter(cls.datasource_name == datasource_name)
+            .filter(DruidCluster.cluster_name == database.name)
```
Comment (john-bodley) [inline superset/connectors/druid/models.py]: Why isn't this `DruidCluster.id == database.id`?

Comment (serenajiang) [reply superset/connectors/druid/models.py]: I am not sure whether database.id has anything to do with cluster.id (it doesn't seem to be related in our database).

Comment (john-bodley) [reply superset/connectors/druid/models.py]: Grokking through the [code](https://github.com/apache/incubator-superset/search?utf8=%E2%9C%93&q=query_datasources_by_name&type=) it seems like this is only called [here](https://github.com/apache/incubator-superset/blob/e6be51953cad5f2872ea8effa40cd1060158f058/superset/security/manager.py#L322) which is SQL specific. It seems like this method should just return an empty list.

@mistercrunch probably has the most context regarding the connection registry.

**Review by john-bodley — APPROVED** [review]

**Pull request merged**